File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Views/Label.kt
package com.abracode.actionui.Views

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.material3.Text as M3Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUIViewConstruction
import com.abracode.actionui.Common.LocalActionUILogger
import com.abracode.actionui.Helpers.LabelIcon
import com.abracode.actionui.Helpers.selectLabelIcon
import com.abracode.actionui.Helpers.stringProperty

/**
 * Renders a title paired with a leading icon - the Android mirror of the Apple
 * `Label` element (`ActionUI/Views/Label.swift`), which maps to SwiftUI `Label`.
 *
 * SwiftUI's `Label(title, systemImage:)` lays an icon before its title; with no
 * image it collapses to `.titleOnly`. This builder does the same: a `Row` of an
 * optional glyph icon and the title text, both inheriting the ambient `font` /
 * `foregroundStyle` (so the icon and text match), with the universal modifiers
 * (padding, frame, background, ...) applied to the row via [modifier].
 *
 * **Icon source.** The icon resolves through the shared `SymbolIcon` seam, the
 * same glyph path as `Image`:
 *   * `systemImage`          -> an SF Symbol, mapped to the closest Material glyph
 *     via the bundled SF->Material map. This is the Apple-canonical key; shared
 *     cross-platform JSON needs nothing else.
 *   * `materialName:android` -> an explicit Material Symbol (the Android escape
 *     hatch), winning over `systemImage` when both are present.
 *   * Axis overrides `materialWeight` / `materialFill` / `materialGrade` /
 *     `materialSize` (`:android`) and `imageScale` tune the glyph, exactly as on
 *     `Image`.
 *
 * **Deferred vs. Apple** (warn-and-skip, not silent): `imageName` (asset-catalog
 * image) maps to an Android `res/drawable` resource, pending the same
 * name->resource contract `Image`'s `assetName` waits on. A `Label` carrying only
 * `imageName` renders title-only with one warning.
 *
 * **Accessibility.** `accessibilityLabel` (and the other universal
 * accessibility properties) arrive on [modifier] via the shared pipeline
 * (`ModifierResolver.applyCommonProperties`); the icon's `contentDescription`
 * stays null so the label is not announced a second time on the glyph node.
 *
 * Sample JSON:
 * ```
 * { "type": "Label", "properties": { "title": "Favorites",
 *   "systemImage": "star.fill", "foregroundStyle": "orange" } }
 * ```
 */
object Label : ActionUIViewConstruction {

    /** Gap between the icon and the title, approximating SwiftUI `Label` spacing. */
    private val IconGap = 6.dp

    @Composable
    override fun BuildView(element: ActionUIElement, modifier: Modifier) {
        val props = element.properties
        val logger = LocalActionUILogger.current

        val title = props?.stringProperty("title") ?: ""
        val imageScale = props?.stringProperty("imageScale")

        // Icon selection (+ its deferred-source warning) runs once per change. A
        // null result is title-only (not an error), unlike Image's no-source case.
        val iconSource = remember(props) { selectLabelIcon(props, "imageName", "Label", logger) }

        Row(
            modifier = modifier,
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(IconGap),
        ) {
            // Shared glyph seam: draws nothing (and warns) on an unknown name, and
            // composes no node when there is no icon, so spacedBy adds no phantom gap.
            LabelIcon(
                source = iconSource,
                imageScale = imageScale,
                contentDescription = null,
                elementName = "Label",
                logger = logger,
            )

            if (title.isNotEmpty()) {
                M3Text(text = title)
            }
        }
    }
}

File: ActionUIWeb/src/Views/Label.js
// Label.js — Label element.
// Web analog of ActionUI/Views/Label.swift (and ActionUIAndroid Views/Label.kt).
//
// A title paired with a leading icon — SwiftUI `Label(title, systemImage:)`,
// collapsing to title-only when there is no image. The icon resolves through the
// shared glyph seam (selectLabelIcon → SymbolIcon.js), the same path as Image and
// Button: `systemImage` (SF Symbol → SF→Material glyph) or `materialName:web`
// (explicit Material glyph, winning when both are present); `imageScale` and the
// `material*` axes tune it. The title is the Label's mutable runtime value
// (valueType String), so get/setElementValue round-trips the text.
//
// Deferred vs. Apple (warn-and-skip, not silent): `imageName` (asset-catalog
// image) needs a name→URL contract web doesn't have yet — the same deferral as
// Image's `assetName` and Button's `assetImage`. A Label carrying only
// `imageName` renders title-only with one warning.

import { register } from "../Common/ActionUIRegistry.js";
import { selectLabelIcon, labelIcon } from "../Helpers/SymbolIcon.js";

register("Label", {
    valueType: "string",

    // Mirrors Label.swift validateProperties (warning text included verbatim).
    validateProperties: (properties, logger) => {
        const validated = { ...properties };
        if (validated.title !== undefined && typeof validated.title !== "string") {
            logger.log("Label title must be a String; ignoring", "warning");
            delete validated.title;
        }
        if (validated.systemImage !== undefined && typeof validated.systemImage !== "string") {
            logger.log("Label systemImage must be a String; ignoring", "warning");
            delete validated.systemImage;
        }
        if (validated.imageName !== undefined && typeof validated.imageName !== "string") {
            logger.log("Label imageName must be a String; ignoring", "warning");
            delete validated.imageName;
        }
        return validated;
    },

    initialValue: (element, properties) => properties.title ?? "",

    buildView: (element, properties, ctx) => {
        const node = document.createElement("span");
        node.className = "aui-label";

        // Leading icon through the shared seam (imageName warns-and-skips). The
        // asset-catalog key is `imageName` on Label, vs `assetImage` on Button.
        const icon = labelIcon(selectLabelIcon(properties, "imageName", "Label", ctx.logger),
            properties, "Label", ctx.logger);
        if (icon) {
            node.classList.add("aui-label-has-icon");
            node.appendChild(icon);
        }

        const titleSpan = document.createElement("span");
        titleSpan.className = "aui-label-title";
        setTitle(titleSpan, properties.title ?? "");
        node.appendChild(titleSpan);

        if (element.id > 0) {
            ctx.model.bind(element.id, {
                getValue: () => titleSpan.textContent,
                setValue: (value) => setTitle(titleSpan, String(value ?? "")),
            });
        }
        return node;
    },
});

// The title is the Label's mutable runtime value. An empty title hides the span
// so the icon→title gap reserves no phantom space (SwiftUI's .titleOnly with an
// empty title is just the icon).
function setTitle(span, text) {
    span.textContent = text;
    span.style.display = text === "" ? "none" : "";
}
